File: ch04/permutate_by_cycle.c
#include <stdlib.h>
#include <stdio.h>

int main(int argc, char **argv)
{
    return 0;
}

int *permuteByCyclic(int *A, int size)
{
    int *B, offset, dest;

    B = malloc(size * sizeof(*B));
    offset = rand() % (size - 1);

    for (int i = 0; i < size; ++i) {
        dest = i + offset;
        if (dest >= size) {
            dest = dest - n + 1;
        }
        B[dest] = A[i];
    }
    
    return B;
}

File: ch05/permutateByCycle.py
import math

def permutateByCyclic(A):
    n = len(A)
    B = [0] * n
    offset = math.rand(0, n-1)
    for i in range(n):
        dest = i + offset
        if dest >= n:
            dest = dest - n + 1
        B[dest] = A[i]
    return B